contrib/pg_tde/src/catalog/tde_principal_key.c

```diff
@@ -881,16 +994,93 @@ pg_tde_is_provider_used(Oid databaseOid, Oid providerId)
 	}
 	else
 	{
-		/* database lcal provider, just verify that it isn't currently active */
+		/* database local provider, just verify that it isn't currently active */
 
-		TDEPrincipalKey *principal_key = GetPrincipalKey(databaseOid, LW_EXCLUSIVE);
+		TDEPrincipalKey *principal_key = GetPrincipalKeyNoDefault(databaseOid, LW_EXCLUSIVE);
 
 		LWLockRelease(tde_lwlock_enc_keys());
 
 		return principal_key != NULL && providerId == principal_key->keyInfo.keyringId;
 	}
 }
 
+static bool
+pg_tde_is_same_principal_key(TDEPrincipalKey *a, TDEPrincipalKey *b)
+{
+	return a != NULL && b != NULL && strncmp(a->keyInfo.keyId.name, b->keyInfo.keyId.name, PRINCIPAL_KEY_NAME_LEN) == 0 && a->keyInfo.keyringId == b->keyInfo.keyringId;
+}
+
+static void
+pg_tde_rotate_default_key_for_database(TDEPrincipalKey *oldKey, TDEPrincipalKey *newKeyTemplate)
+{
+	bool		is_rotated;
+
+	TDEPrincipalKey *newKey = palloc(sizeof(TDEPrincipalKey));
+
+	memcpy(newKey, newKeyTemplate, sizeof(TDEPrincipalKey));
+	newKey->keyInfo.databaseId = oldKey->keyInfo.databaseId;
+
+	/* key rotation */
+	is_rotated = pg_tde_perform_rotate_key(newKey, oldKey);
+
+	if (is_rotated && (!TDEisInGlobalSpace(newKey->keyInfo.databaseId)))
+	{
+		clear_principal_key_cache(oldKey->keyInfo.databaseId);
+		push_principal_key_to_cache(newKey);
+	}
+
+	pfree(newKey);
+}
+
+static void
+pg_tde_update_global_principal_key_everywhere(TDEPrincipalKey *oldKey, TDEPrincipalKey *newKey)
+{
+	HeapTuple	tuple;
+	SysScanDesc scan;
+	Relation	rel;
+	Oid			dbOid = GLOBAL_DATA_TDE_OID;
+	TDEPrincipalKey *principal_key;
+
+	/* First check the global oid */
+	principal_key = GetPrincipalKeyNoDefault(dbOid, LW_EXCLUSIVE);
+
+	if (pg_tde_is_same_principal_key(oldKey, principal_key))
+	{
+		pg_tde_rotate_default_key_for_database(principal_key, newKey);
+	}
+
+
+	/*
+	 * Take exclusive lock, as we do not want anybody to create/drop a
+	 * database in parallel
+	 */
+	/* If it happens, its not the end of the world, but not ideal */
+	rel = table_open(DatabaseRelationId, AccessExclusiveLock);
+
+	scan = systable_beginscan(rel, 0, false, NULL, 0, NULL);
+
+	while ((tuple = systable_getnext(scan)) != NULL)
+	{
+		if (!HeapTupleIsValid(tuple))
+		{
+			break;
+		}
```

Comment:
Why this check? All tuples should be valid, right? And if not I do not think breaking from the loop makes much sense.

Reply:
Right, this just repeats the loop condition. At first I just copy-pasted the same loop from the core, but that was an infinite loop using this as the exit. I refactored it to a proper while loop, and forgot this here.